### PR TITLE
Fix permanent CDC event loss caused by premature PostgreSQL LSN acknowledgment

### PR DESCRIPTION
**Pull request by WHOIM1205**
## Description

This PR fixes a **critical CDC correctness bug** in the PostgreSQL source where WAL offsets (LSNs) were acknowledged **before events were durably delivered** to downstream consumers.

Previously, the replication stream advanced and reported the LSN to PostgreSQL immediately upon receiving `XLogData`, even if:
- WAL decoding failed
- Event dispatch failed
- Events were dropped due to backpressure
- The source pod crashed before delivery completed

Because PostgreSQL relies on replication feedback to decide which WAL segments can be discarded, this behavior could lead to **permanent and silent data loss** in real-world scenarios.

### What this PR changes
- Introduces a separate `confirmed_lsn` to track WAL positions that have been **successfully dispatched**
- Ensures PostgreSQL `flush_lsn` and `apply_lsn` are only advanced **after successful delivery**
- Makes dispatch failures **fatal** to prevent silent offset advancement
- Preserves existing behavior while restoring CDC safety guarantees

---

## Type of change

- [x] **BUGFIX** (non-breaking change which fixes a critical issue)

---

### Tests Added
- Added unit tests covering:
  - LSN parsing and initialization
  - Transaction lifecycle behavior
  - `confirmed_lsn` not advancing on `BEGIN`
  - `confirmed_lsn` advancing only after successful `COMMIT`
  - Replication feedback using `confirmed_lsn` instead of `current_lsn`
  - Safety invariants when transactions are pending

### Verified locally with
```bash
cargo test -p drasi-source-postgres

### CODE REVIEW

**Review by danielgerlag — COMMENTED** [review]
## Review: LSN Acknowledgment Implementation

Hi @WHOIM1205! 👋

Thank you for taking the time to contribute to Drasi and for thinking deeply about CDC correctness guarantees. The core insight here—that `confirmed_lsn` should be separate from `current_lsn` and only advance after successful dispatch—is absolutely correct for scenarios involving durable downstream delivery. The documentation you've added explaining the LSN semantics is also excellent and will help future maintainers.

I've reviewed the implementation in detail and have some feedback that I hope will be helpful. Please don't be discouraged by the length—this is a nuanced area and your approach shows good understanding of the problem space.

---

### 🔴 Issue 1: Empty Dispatchers Treated as Fatal Error (High Impact)

**Location:** `dispatch_event_directly()` method

```rust
if dispatchers_guard.is_empty() {
    return Err(anyhow!(
        "[{}] No dispatchers registered - cannot confirm LSN without delivery",
        self.source_id
    ));
}
```

**Concern:** This will cause the source to fail repeatedly during normal operational scenarios:

- **Startup timing:** The source may start before any queries have subscribed
- **Reconfiguration:** Queries may be temporarily removed and re-added
- **Dynamic subscriptions:** Drasi supports adding/removing queries at runtime

With this change, the source would crash-loop until a subscriber appears, which could prevent the system from ever reaching a stable state.

**Suggestion:** Consider one of these alternatives:
1. Buffer events temporarily until the first subscriber appears
2. Add a configuration option for the behavior (`fail_on_no_subscribers: bool`)
3. Log a warning but treat "no subscribers" as "vacuously delivered" for in-memory scenarios

---

### 🟡 Issue 2: First Dispatcher Failure Aborts Entire Multi-Subscriber Dispatch (Medium Impact)

**Location:** `dispatch_event_directly()` method

```rust
for dispatcher in dispatchers_guard.iter() {
    if let Err(e) = dispatcher.dispatch_change(arc_wrapper.clone()).await {
        return Err(anyhow!("Dispatcher failed to deliver event: {}", e));
    }
}
```

**Concern:** If there are multiple subscribers (dispatchers) and the first one fails, the remaining dispatchers never receive the event. This creates an inconsistency where some subscribers are behind others.

For durable dispatch scenarios, you're right that we need "all or nothing" semantics. However, the current implementation doesn't roll back the dispatchers that already received the event before the failure.

**Suggestion:** Consider:
1. Collecting all dispatch results first, then deciding whether to confirm
2. Adding a dispatch strategy configuration (`AllOrNothing`, `BestEffort`, `RequireQuorum(n)`)
3. For true atomicity, you'd need a two-phase approach (prepare/commit)

---

### 🟢 Issue 3: Decode Errors Can Block Progress Indefinitely (Lower Impact)

**Location:** `handle_xlog_data()` error handling

```rust
Err(e) => {
    debug!("Failed to decode WAL message...");
    // Don't advance confirmed_lsn on decode errors
}
```

**Concern:** If a WAL message is persistently undecodable (corruption, unsupported PostgreSQL version feature, etc.), the source will:
1. Fail to decode
2. Not advance `confirmed_lsn`
3. Receive the same message on next feedback cycle
4. Loop forever

While this is conservative and safe, it could wedge the source permanently.

**Suggestion:** Consider adding a circuit breaker:

```rust
// After N failures for the same LSN, skip with a warning
if self.decode_failure_count.get(&end_lsn).unwrap_or(&0) > MAX_DECODE_RETRIES {
    warn!("Skipping undecodable message at LSN {:x} after {} retries", end_lsn, MAX_DECODE_RETRIES);
    self.confirmed_lsn = end_lsn;
    return Ok(());
}
```

---

### 📝 Missing Test Coverage

The tests you've added for the happy path are excellent! To make this change production-ready, I'd recommend adding tests for error scenarios:

| Scenario | Why It Matters |
|----------|----------------|
| `test_dispatch_failure_does_not_confirm_lsn` | Verifies the core safety guarantee |
| `test_empty_dispatchers_returns_error` | Documents the current behavior |
| `test_partial_transaction_rollback_on_failure` | Ensures mid-transaction failures are handled |
| `test_recovery_after_dispatch_failure` | Confirms the source can resume |
| `test_decode_error_handling` | Covers the decode failure path |

---

### Summary

| Aspect | Assessment |
|--------|------------|
| Core approach | ✅ Correct for durable dispatch scenarios |
| LSN separation | ✅ Well-designed |
| Documentation | ✅ Excellent |
| Error handling | ⚠️ Needs refinement |
| Test coverage | ⚠️ Missing error paths |

I think this PR has good bones and addresses a real concern for future durable dispatch scenarios. With some refinements to the error handling and additional test coverage, it could be a solid addition.

Happy to discuss any of these points further! 🙂

File: components/sources/postgres/src/stream.rs [hunk 1]
```diff
@@ -787,3 +898,357 @@ fn parse_lsn(lsn_str: &str) -> Result<u64> {
...
+            flush_lsn: 0x3000,  // What we dispatched
+            apply_lsn: 0x3000,  // What downstream processed
+            reply_requested: false,
+        };
+
+        // The fix ensures flush_lsn can be different from write_lsn
+        assert!(status.write_lsn > status.flush_lsn);
+        assert_eq!(status.flush_lsn, status.apply_lsn);
+    }
+}
```
[the remark this answers is not in the document]
Comment (WHOIM1205) [reply components/sources/postgres/src/stream.rs:1377]: Thanks for the detailed review.
This PR intentionally focuses on fixing the core correctness issue: preventing premature LSN acknowledgment when dispatch cannot be guaranteed. Existing tests already assert the key safety invariant that confirmed_lsn only advances on successful paths.
The additional edge cases you mentioned are valid, but addressing them would expand the scope of this fix. I’m happy to follow up with additional targeted tests in a separate PR if needed.

[the remark this answers is not in the document]
Comment (danielgerlag) [reply components/sources/postgres/src/stream.rs:1377]: Lets add these tests

**Review by danielgerlag — CHANGES_REQUESTED** [review]
Hi @WHOIM1205 👋

   Thank you for your continued work on this — the core insight behind separating `confirmed_lsn` from `current_lsn` is exactly right, and the direct dispatch
  approach that bypasses `SourceBase::dispatch_from_task` is a smart way to get proper error propagation without breaking other sources. The documentation you've
   added explaining the LSN semantics is also really helpful.

   I've done another pass and wanted to share where things stand so you can decide if you'd like to continue working on this. No pressure at all — I understand
  this is a tricky area and I appreciate the time you've already invested.

   ---

   ### Items that still need attention

   #### 1. Keepalive can still advance `confirmed_lsn` past unprocessed records (Critical)

   In both `handle_keepalive` (line ~384) and `handle_message` for `PrimaryKeepaliveMessage` (line ~237), `confirmed_lsn` is set to `wal_end` when there's no 
  pending transaction. The problem is that `wal_end` from a keepalive reflects PostgreSQL's *current write position*, which can be ahead of the last record we
  actually decoded and dispatched.

   Consider this sequence:
   1. A WAL record arrives → decode fails or returns `Ok(None)` → `confirmed_lsn` correctly stays put
   2. A keepalive arrives with `wal_end` past that record → `confirmed_lsn` jumps forward
   3. The unprocessed record is now acknowledged and PostgreSQL may discard it

   The fix would be to only update `current_lsn` in keepalive handlers, not `confirmed_lsn`. Keepalives should just trigger sending feedback using the *existing*
   `confirmed_lsn`.

   #### 2. `end_lsn` from XLogData header may overshoot (Medium)

   The `end_lsn` extracted from the XLogData header (bytes 8–15) is the server's WAL end position at send time, not necessarily the end of the payload in *this*
  message. Using it as `message_lsn` for non-transactional changes could acknowledge past the current record. For transactional commits this is fine since you 
  use `tx_info.commit_lsn` — but the non-transactional path (`dispatch_single_change`) passes this potentially-ahead value.

   #### 3. `warn!` log level for no-dispatcher case (Minor)

   `dispatch_event_directly` logs at `warn!` for every event when no dispatchers are registered. During normal startup (source starts before queries subscribe), 
  this could produce a lot of noise. `debug!` or `trace!` would be more appropriate here.

   #### 4. Additional test coverage

   In my earlier review I asked for tests covering a few specific scenarios. The `test_dispatch_failure_does_not_advance_confirmed_lsn` test you added is great —
   that's the most important one. The remaining scenarios I'd still like to see covered are:

   - Empty dispatcher list behavior
   - Decoder returning `Ok(None)` — LSN should not advance
   - Truncate operation handling
   - `test_feedback_uses_confirmed_lsn_not_current_lsn` currently only checks initial state (same as `test_initial_lsn_state`) — it should verify the feedback
  payload actually uses `confirmed_lsn`

   #### 5. Unused import

   `BroadcastChangeDispatcher` is imported but not used in the test module (line ~920).

   ---

   ### What's working well ✅

   - The `confirmed_lsn` / `current_lsn` separation and using `confirmed_lsn` for `flush_lsn`/`apply_lsn` in feedback
   - Direct dispatch with proper error propagation
   - Attempting all dispatchers before deciding failure (addresses the multi-subscriber concern)
   - Vacuous delivery for no-dispatcher case (handles startup ordering gracefully)
   - Empty transaction LSN confirmation
   - The `FailingDispatcher` mock test proving the core safety invariant

   ---

   If you're still interested in continuing with this PR, I'm happy to answer any questions or discuss the keepalive issue in more detail. If your availability
  has changed, that's completely fine too — just let me know and we can figure out next steps. Thanks again for the contribution!

**Comment by WHOIM1205** [comment]
> Hi @WHOIM1205 👋
> 
>    Thank you for your continued work on this — the core insight behind separating `confirmed_lsn` from `current_lsn` is exactly right, and the direct dispatch
>   approach that bypasses `SourceBase::dispatch_from_task` is a smart way to get proper error propagation without breaking other sources. The documentation you've
>    added explaining the LSN semantics is also really helpful.
> 
>    I've done another pass and wanted to share where things stand so you can decide if you'd like to continue working on this. No pressure at all — I understand
>   this is a tricky area and I appreciate the time you've already invested.
> 
>    ---
> 
>    ### Items that still need attention
> 
>    #### 1. Keepalive can still advance `confirmed_lsn` past unprocessed records (Critical)
> 
>    In both `handle_keepalive` (line ~384) and `handle_message` for `PrimaryKeepaliveMessage` (line ~237), `confirmed_lsn` is set to `wal_end` when there's no 
>   pending transaction. The problem is that `wal_end` from a keepalive reflects PostgreSQL's *current write position*, which can be ahead of the last record we
>   actually decoded and dispatched.
> 
>    Consider this sequence:
>    1. A WAL record arrives → decode fails or returns `Ok(None)` → `confirmed_lsn` correctly stays put
>    2. A keepalive arrives with `wal_end` past that record → `confirmed_lsn` jumps forward
>    3. The unprocessed record is now acknowledged and PostgreSQL may discard it
> 
>    The fix would be to only update `current_lsn` in keepalive handlers, not `confirmed_lsn`. Keepalives should just trigger sending feedback using the *existing*
>    `confirmed_lsn`.
> 
>    #### 2. `end_lsn` from XLogData header may overshoot (Medium)
> 
>    The `end_lsn` extracted from the XLogData header (bytes 8–15) is the server's WAL end position at send time, not necessarily the end of the payload in *this*
>   message. Using it as `message_lsn` for non-transactional changes could acknowledge past the current record. For transactional commits this is fine since you 
>   use `tx_info.commit_lsn` — but the non-transactional path (`dispatch_single_change`) passes this potentially-ahead value.
> 
>    #### 3. `warn!` log level for no-dispatcher case (Minor)
> 
>    `dispatch_event_directly` logs at `warn!` for every event when no dispatchers are registered. During normal startup (source starts before queries subscribe), 
>   this could produce a lot of noise. `debug!` or `trace!` would be more appropriate here.
> 
>    #### 4. Additional test coverage
> 
>    In my earlier review I asked for tests covering a few specific scenarios. The `test_dispatch_failure_does_not_advance_confirmed_lsn` test you added is great —
>    that's the most important one. The remaining scenarios I'd still like to see covered are:
> 
>    - Empty dispatcher list behavior
>    - Decoder returning `Ok(None)` — LSN should not advance
>    - Truncate operation handling
>    - `test_feedback_uses_confirmed_lsn_not_current_lsn` currently only checks initial state (same as `test_initial_lsn_state`) — it should verify the feedback
>   payload actually uses `confirmed_lsn`
> 
>    #### 5. Unused import
> 
>    `BroadcastChangeDispatcher` is imported but not used in the test module (line ~920).
> 
>    ---
> 
>    ### What's working well ✅
> 
>    - The `confirmed_lsn` / `current_lsn` separation and using `confirmed_lsn` for `flush_lsn`/`apply_lsn` in feedback
>    - Direct dispatch with proper error propagation
>    - Attempting all dispatchers before deciding failure (addresses the multi-subscriber concern)
>    - Vacuous delivery for no-dispatcher case (handles startup ordering gracefully)
>    - Empty transaction LSN confirmation
>    - The `FailingDispatcher` mock test proving the core safety invariant
> 
>    ---
> 
>    If you're still interested in continuing with this PR, I'm happy to answer any questions or discuss the keepalive issue in more detail. If your availability
>   has changed, that's completely fine too — just let me know and we can figure out next steps. Thanks again for the contribution!

Yeah sure i will fix this by tommorow

**Review by danielgerlag — CHANGES_REQUESTED** [review]
Hi @WHOIM1205, thanks for your continued work on this — the core insight behind separating `confirmed_lsn` from `current_lsn` is exactly right, and the direct dispatch approach that bypasses `SourceBase::dispatch_from_task` is a smart way to get proper error propagation without breaking other sources.

I've done another pass and there are a few things that still need attention before we can merge. Most importantly, there are paths in the current code that can still advance `confirmed_lsn` prematurely, which undermines the safety guarantee this PR is trying to establish.

---

### Critical: The fix is still incomplete in a few spots

**1. Keepalive handlers still advance `confirmed_lsn` past unprocessed records**

Both `handle_keepalive` (line ~382) and the `PrimaryKeepaliveMessage` branch in `handle_message` (line ~244) set `confirmed_lsn = wal_end` when no transaction is pending. The problem is that `wal_end` from a keepalive reflects PostgreSQL's *current write position*, which can be ahead of the last record we actually decoded and dispatched.

Consider this scenario:
1. A WAL record arrives → decode fails or returns `Ok(None)` → `confirmed_lsn` correctly stays put
2. A keepalive arrives with `wal_end` past that record → `confirmed_lsn` jumps forward
3. The unprocessed record is now acknowledged and PostgreSQL may discard it

The fix is straightforward: keepalive handlers should only update `current_lsn`, not `confirmed_lsn`. They should just trigger sending feedback using the *existing* `confirmed_lsn`. (I also noticed the comment on line ~377 says "NOTE: Don't update confirmed_lsn here" but then does exactly that a few lines later — the comment and code are contradicting each other.)

**2. `end_lsn` from XLogData header may overshoot for non-transactional changes**

For non-transactional inserts/updates/deletes, `message_lsn` (from the XLogData header bytes 8–15) is used to advance `confirmed_lsn` via `dispatch_single_change`. But this field is the server's WAL end position at send time, not necessarily the end of the payload in *this* message. The transactional path correctly uses `tx_info.commit_lsn` — the non-transactional path should use a similarly reliable value.

**3. Relation handler advances `confirmed_lsn` unconditionally during pending transactions** (line ~474)

`WalMessage::Relation` sets `confirmed_lsn = message_lsn` without checking `self.pending_transaction.is_none()`. Every other message type (Insert/Update/Delete) guards this. If a Relation message arrives mid-transaction (e.g., from an ALTER TABLE), `confirmed_lsn` would jump ahead before the transaction commits.

---

### Previously requested changes still outstanding

**4. Additional test coverage**

In my earlier review I asked for tests covering these specific scenarios, and I'd still like to see them in this PR:
- Empty dispatcher list behavior
- Decoder returning `Ok(None)` — LSN should not advance
- Truncate operation handling
- `test_feedback_uses_confirmed_lsn_not_current_lsn` currently only checks that both LSNs start at 0 (same as `test_initial_lsn_state`) — it should verify the feedback payload actually uses `confirmed_lsn`

**5. `warn!` log level for no-dispatcher case** (`dispatch_event_directly`, line ~601)

During normal startup, sources often start before queries subscribe, which means every event would log at `warn!` level. This could produce a lot of noise. `debug!` or `trace!` would be more appropriate.

---

### Housekeeping

- **Merge conflicts**: The PR has conflicts with `main` that need to be resolved.
- **Unused import**: `BroadcastChangeDispatcher` is imported in the test module but never used.
- **No linked issue**: Per our [contributing guidelines](https://github.com/drasi-project/drasi-core/blob/main/CONTRIBUTING.md), please link the relevant issue using "Fixes #\<issue-number\>" in the PR description.

---

### What's working well ✅

- The `confirmed_lsn` / `current_lsn` separation and using `confirmed_lsn` for `flush_lsn`/`apply_lsn` in feedback
- Direct dispatch with proper error propagation
- Attempting all dispatchers before deciding failure
- Vacuous delivery for no-dispatcher case (handles startup ordering gracefully)
- The `FailingDispatcher` mock test proving the core safety invariant
- Transaction workflow test with real `ChannelChangeDispatcher`
- Empty transaction LSN confirmation
- Clear documentation explaining the LSN semantics

---

I understand this is a tricky area and I appreciate the time you've invested. The approach is sound — it just needs these remaining gaps closed so we don't ship a partial fix for a data loss bug. Let me know if you have any questions or if your availability has changed and we need to figure out next steps. Thanks!
